Invoke the locale plugin rake tasks directly

jrafanie opened the pull request:
We have lots of plugins and there should be no need to shell out and constantly
load the rails environment and application over and over again.

I've verified that we're generating the same po and pot files with direct rake
task invocation but we save several minutes.

Before this commit:

```
bundle exec rake locale:update_all  248.95s user 83.08s system 96% cpu 5:43.01 total
```

After this commit:

```
bundle exec rake locale:update_all  78.93s user 8.90s system 97% cpu 1:29.70 total
```

Here is a book describing how I tested this:

Po/pot files are not updated each time we change strings throughout our codebase so they become stale quickly.  Therefore we need to generate a baseline on master before we can evaluate if any changes we make to the po / pot generation, for refacoring or performance improvements, accidentally changed the outputted files.

Here's how I verify this.

### Capture the baseline message catalogs on master

1) git checkout master
2) git checkout -b baseline
3) bundle exec rake locale:update_all # generates the updated message catalogs for master
4) git add locale
5) git commit -m "Add baseline master catalogs"
6) Do step 3 again and check git diff # this should show only a few timestamp fields updated

### Start a "changes" branch

1) git checkout master # use the same base branch as the baseline to avoid plugin/other changes
2) git checkout -b "changes" # Now, make your code changes
3) bundle exec rake locale:update_all # verify locale/ has changes to the po/pot
4) git add locale
5) git commit -m "my changes"
6) Compare the locale/ directory from this new branch to the baseline:
6b) git diff baseline HEAD locale/ # this should show only a few timestamp fields updated
7) Do step 3 again # this should show only a few timestamp fields updated

If you are just refactoring code or making performance improvements, 6b should show no differences from the baseline.

Fryguy commented on lib/tasks/locale.rake:
```diff
@@ -175,8 +175,8 @@ namespace :locale do
     Vmdb::Plugins.each do |plugin|
-      system('bundle', 'exec', 'rake', "locale:delete_pot_file[#{plugin.root}]") # Delete plugin's pot file if it exists to avoid weird file timestamp issues
-      raise unless system('bundle', 'exec', 'rake', "locale:plugin:find[#{plugin.to_s.sub('::Engine', '')}]")
+      Rake::Task['locale:delete_pot_file'].invoke(plugin.root) # Delete plugin's pot file if it exists to avoid weird file timestamp issues
+      Rake::Task['locale:plugin:find'].invoke(plugin.to_s.sub('::Engine', '')) # will warn and exit 1 if any engine fails
```
Since this is run for each plugin, does it actually get run or does Rake "skip" it once it's been run at least once

jrafanie replied:
Ah good point... rake tasks won't be invoked a section time.  Let me check if different arguments change that.

jrafanie replied:
The time appears to stay the same... about 1 minute.  I'm currently checking if the contents are different vs. master.

jrafanie replied:
> Since this is run for each plugin, does it actually get run or does Rake "skip" it once it's been run at least once. Put another way, what do you see when you run it with `--trace`

Great find.  It's different now that I have all the plugins running and it looks like the global shared state is why this code was previously invoking a new shell session for each command.  Making this WIP.

jrafanie replied:
Ok, it took a while to test by generating on master and doing it over and generating on this branch to confirm that the files generated are the same.

This is the little hack that allows us to avoid creating new processes to run rake tasks in the context of each plugin:

```ruby
Rake.application.tasks.each {|t| t.reenable if (t.already_invoked && !t.kind_of?(Rake::FileTask))}
```

jrafanie replied:
The solution should be to change some of these rake tasks to method so we can call the method multiples times in the same rake tasks and not skip it.  Additionally, we can avoid loading environment if we have to go back to using extra process invocations but I think this solution is simple enough until we can fix the underlying rake tasks so they can cleanup after themselves and not be skipped when run for each plugin.

jrafanie commented:
Ok, I've simplified this PR down to the basics and added a 📖 on how I tested this to ensure I didn't break how we generate message catalogs.

Basically, this branch and master will generate the same message catalogs with `rake locale:update_all` but this branch will take over 4 minutes left.
